tfworker/util/copier.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2021 Richard Maynard (richard.maynard@gmail.com)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.import collections
+import os
+import re
+import shutil
+import tempfile
+from abc import ABCMeta, abstractmethod, abstractstaticmethod
+from pathlib import Path
+from typing import Callable
+
+from tfworker.util.system import pipe_exec
+
+
+class CopyFactory:
+    """ The factory class for creating copiers """
+
+    registry = {}
+
+    @classmethod
+    def register(cls, name: str) -> Callable:
+        """ Class method to register copiers """
+
+        def inner_wrapper(wrapped_class: Copier) -> Callable:
+            if name in cls.registry:
+                raise ValueError(f"Executor {name} already exists")
+            cls.registry[name] = wrapped_class
+            return wrapped_class
+
+        return inner_wrapper
+
+    @classmethod
+    def create(cls, source: str, **kwargs) -> "Copier":
+        """ create returns a copier type that supports handling the provided source """
+        copier_class = cls.registry[cls.get_copier_type(source, **kwargs)]
+        copier = copier_class(source, **kwargs)
+        return copier
+
+    @classmethod
+    def get_copier_type(cls, source: str, **kwargs) -> str:
+        """ get_copier_type tries to find a supported copier based on the provided source """
+        for copier_type, copier_class in cls.registry.items():
+            if copier_class.type_match(source, **kwargs):
+                return copier_type
+        raise NotImplementedError(f"no valid copier for {source}")
+
+
+class Copier(metaclass=ABCMeta):
+    """ The base class for definition copiers """
+
+    def __init__(self, source: str, **kwargs):
+        self._source = source
+
+        for k, v in kwargs.items():
+            setattr(self, f"_{k}", v)
```

Comment:
This might lead to some unexpected behavior later on. If you were to add a private method beginning in `_` I'm not sure what would happen if you also passed in a kwarg of the same name:
```
class NewCoper(...):

    def _check(self):
        ...
```
NewCopier(check=False)

I'm not sure if `NewCopier()._check` would refer to the method or the attribute.

Reply:
That's a good point, I originally was checking for specific values. I could revert to that, or perhaps I could do a check with hasattr to make sure there's no conflicts